### ISSUE TITLE
new rule - report an error on `Divergent` types

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3756]
as far as i can tell, `Divergent` is treated the same as `Any`, so ideally it should be an error (related: #622):

```py
from typing import reveal_type

def some_condition() -> bool:
    ...

x = [1]
while some_condition():
  x = [x]

value = x[0]

if not isinstance(value, int):
    value.asdf # no error
```

[Playground](https://play.ty.dev/49bd16c4-7b67-4fde-b9fa-d3adddb444f5)

there also doesn't seem to be any docs on what `Divergent` means. if it's the same as `Unknown`/`Any`, i would expect it to be documented [here](https://docs.astral.sh/ty/reference/typing-faq/#what-is-the-unknown-type-and-when-does-it-appear)

pyrefly's [`non-convergent-recursion`](https://pyrefly.org/en/docs/error-kinds/#non-convergent-recursion) rule catches this

**Comment by AlexWaygood** [comment]
Arguably any time `Divergent` appears in a user-visible type, that's a ty limitation that we should aim to fix :-)

**Comment by ibraheemdev** [comment]
> Arguably any time `Divergent` appears in a user-visible type, that's a ty limitation that we should aim to fix :-)

Hmm... I think `Divergent` can be useful signal to users for catching infinite recursion. What would you expect the type of `x` to be here?
```py
def flag() -> bool: ...

x = []
while flag():
  x.append(x)

reveal_type(x)  # revealed: list[Divergent]
```

**Comment by carljm** [comment]
I agree with @ibraheemdev -- `Divergent` is often inherent to the types, not a ty inference limitation. And I'm not sure it's better for users if we just flatten it to `Unknown` vs explicitly signaling that there is non-convergent recursion happening. In other words, I don't consider appearance of `Divergent` in a user-visible type to represent a ty limitation we should fix.

I do think having a diagnostic in addition to the `Divergent` type, and documenting `Divergent`, would both make sense.

**Comment by AlexWaygood** [comment]
Makes sense

**Comment by carljm** [comment]
I've put up a PR for the documentation piece; I think the diagnostic is a nice-to-have for stable release, so putting this issue in 1.1 milestone.

**Comment by sharkdp** [comment]
> I agree with [@ibraheemdev](https://github.com/ibraheemdev) -- `Divergent` is often inherent to the types, not a ty inference limitation.

In principle, we *could* infer a more precise type here (`type Rec = list[Rec]`). That would catch errors like `x[0].whatever`. I'm not sure if that's (even theoretically) possible to implement, though. But in that sense, I understand where Alex's "ty limitation" comment comes from. `Divergent` does seem like a bit of an escape hatch that avoids having to "close the loop".
